stdlib/camlinternalQuote.ml

```diff
@@ -1196,502 +1292,781 @@ module Ast = struct
+        pat
+    in
+    mk (For (pat, begin_, end_, dir, body)) [Loop]
+
+  let for_simple loc name begin_ end_ dir body =
+    let dir = if dir then Ast.Upto else Ast.Downto in
+    let+ var, body =
+      With_free_vars.value_binding loc name (fun var ->
+          let+ body = body var in
+          var, body)
+    and+ begin_ = begin_
+    and+ end_ = end_ in
+    mk (For (Var var, begin_, end_, dir, body)) [Loop]
+
+  let send exp meth =
+    let+ exp = exp in
+    mk (Ast.Send (exp, meth)) []
+
+  let constraint_ exp constr =
+    let+ exp = exp and+ constr = constr in
+    let desc =
+      match constr with
+      | Ast.Constraint typ -> Ast.ConstraintExp (exp, typ)
+      | Ast.Coerce (typ1, typ2) -> Ast.CoerceExp (exp, typ1, typ2)
+    in
+    mk desc []
+
+  let assert_ exp =
+    let+ exp = exp in
+    mk (Assert exp) []
+
+  let lazy_ exp =
+    let+ exp = exp in
+    mk (Lazy exp) []
+
+  let open_ override m rhs =
+    let flag = if override then Ast.Override else Ast.Fresh in
+    let+ exp = rhs
+    and+ m = m in
+    mk (OpenExp (flag, m, exp)) []
+
+  let letmodule loc name_opt mod_exp body =
+    let+ mod_exp = mod_exp
+    and+ (var, body) =
+      match name_opt with
+      | None ->
+        let+ body = body None
+        in (None, body)
+      | Some name ->
+        With_free_vars.module_binding loc name
+          (fun m ->
+             let+ body = body (Some m)
+             in (Some m, body))
+    in
+    mk (Letmodule (var, mod_exp, body)) []
+
+  let constraint_ exp constr =
+    let+ exp = exp
+    and+ constr = constr
+    in match constr with
+    | Ast.Coerce (ty_opt, ty_to) -> mk (Ast.CoerceExp (exp, ty_opt, ty_to)) []
+    | Ast.Constraint ty -> mk (Ast.ConstraintExp (exp, ty)) []
+
+  let new_ class_id = return (mk (Ast.New class_id) [])
+
+  let pack m =
+    let+ m = m in
+    mk (Ast.Pack m) []
+
+  let unreachable = return (mk Ast.Unreachable [])
+
+  let src_pos = return (mk Ast.Src_pos [])
+
+  let exclave exp =
+    let+ exp = exp
+    in mk (Ast.Exclave exp) []
+
+  let list_comprehension compr =
+    let+ compr = compr
+    in mk (Ast.List_comprehension compr) []
+
+  let array_comprehension compr =
+    let+ compr = compr
+    in mk (Ast.Array_comprehension compr) []
+
+  let unboxed_tuple fs =
+    let entries =
+      List.map
+        (fun (lab, exp) ->
+           let+ exp = exp in
+           lab, exp)
+        fs
+    in
+    let+ entries = all entries in
+    mk (Ast.Unboxed_tuple entries) []
+
+  let unboxed_record_product entries exp =
+    let entries =
+      List.map
+        (fun (field, e) ->
+           let+ e = e in
+           field, e)
+        entries
+    in
+    let+ entries = all entries and+ exp = optional exp in
+    mk (Ast.Unboxed_record_product (entries, exp)) []
+
+  let unboxed_field exp field =
+    let+ exp = exp in
+    mk (Ast.Unboxed_field (exp, field)) []
+
+  let extension_constructor ident =
+    return (mk (Ast.Extension_constructor ident) [])
```

Comment:
Identifiers should always have their free variables included using one of the `with_free_vars` functions. Possibly we should change the API on `Identifier` to try to ensure that this always happens.